src/main/scala/firrtl/Utils.scala

```diff
@@ -569,6 +569,81 @@ object Utils {
        case _ => NoInfo
     }}
 
+  /** Splits an Expression into root Ref and tail
+    *
+    * @example
+    *   Given:   SubField(SubIndex(SubField(Ref("a", UIntType(IntWidth(32))), "b"), 2), "c")
+    *   Returns: (Ref("a"), SubField(SubIndex(Ref("b"), 2), "c"))
+    *   a.b[2].c -> (a, b[2].c)
+    * @example
+    *   Given:   SubField(SubIndex(Ref("b"), 2), "c")
+    *   Returns: (Ref("b"), SubField(SubIndex(EmptyExpression, 2), "c"))
+    *   b[2].c -> (b, EMPTY[2].c)
+    * @note This function only supports WRef, WSubField, and WSubIndex
+    */
+  def splitRef(e: Expression): (WRef, Expression) = e match {
+    case e: WRef => (e, EmptyExpression)
+    case e: WSubIndex =>
+      val (root, tail) = splitRef(e.exp)
+      (root, WSubIndex(tail, e.value, e.tpe, e.gender))
+    case e: WSubField =>
+      val (root, tail) = splitRef(e.exp)
+      tail match {
+        case EmptyExpression => (root, WRef(e.name, e.tpe, root.kind, e.gender))
+        case exp => (root, WSubField(tail, e.name, e.tpe, e.gender))
+      }
+  }
+
+  /** Adds a root reference to some SubField/SubIndex chain */
+  def mergeRef(root: WRef, body: Expression): Expression = body match {
+    case e: WRef =>
+      WSubField(root, e.name, e.tpe, e.gender)
+    case e: WSubIndex =>
+      WSubIndex(mergeRef(root, e.exp), e.value, e.tpe, e.gender)
+    case e: WSubField =>
+      WSubField(mergeRef(root, e.exp), e.name, e.tpe, e.gender)
+    case EmptyExpression => root
+  }
+
+  case class DeclarationNotFoundException(msg: String) extends FIRRTLException(msg)
+
+  /** Gets the root declaration of an expression
+    *
+    * @param m the [[firrtl.InModule]] to search
+    * @param expr the [[firrtl.Expression]] that refers to some declaration
+    * @return the [[firrtl.IsDeclaration]] of `expr`
+    * @throws DeclarationNotFoundException if no declaration of `expr` is found
+    */
+  def getDeclaration(m: InModule, expr: Expression): IsDeclaration = {
```

Comment:
We really need to restructure IR vs WIR, because its starting to bleed everywhere - e.g. which Util functions are for WIR vs IR?